api/scaler/dispatcher.go

```diff
@@ -2,85 +2,60 @@ package scaler
 
 import (
 	"log"
-	"sync"
 	"time"
 
-	"github.com/quintilesims/layer0/api/provider"
-)
-
-const (
-	SCALE_GRACE_PERIOD = time.Second * 15
-)
-
-var (
-	timeMultiplier time.Duration = 1
+	"github.com/quintilesims/layer0/api/job"
+	"github.com/quintilesims/layer0/common/models"
 )
 
 type Dispatcher struct {
-	environmentProvider provider.EnvironmentProvider
-	scaler              Scaler
-	schedule            map[string]*time.Timer
-	lock                *sync.Mutex
+	jobStore    job.Store
+	gracePeriod time.Duration
+	scheduleOps chan func(map[string]*time.Timer)
 }
 
-func NewDispatcher(e provider.EnvironmentProvider, s Scaler) *Dispatcher {
-	return &Dispatcher{
-		environmentProvider: e,
-		scaler:              s,
-		schedule:            map[string]*time.Timer{},
-		lock:                &sync.Mutex{},
+func NewDispatcher(jobStore job.Store, gracePeriod time.Duration) *Dispatcher {
+	d := &Dispatcher{
+		jobStore:    jobStore,
+		gracePeriod: gracePeriod,
+		scheduleOps: make(chan func(map[string]*time.Timer)),
 	}
-}
 
-func (d *Dispatcher) ScheduleRun(environmentID string) {
-	d.lock.Lock()
-	defer d.lock.Unlock()
+	go d.loop()
+	return d
+}
 
-	if timer, ok := d.schedule[environmentID]; ok {
-		log.Printf("[DEBUG] [ScalerDispatcher] Pushing back scheduled run for environment %s", environmentID)
-		timer.Stop()
+func (d *Dispatcher) loop() {
+	schedule := map[string]*time.Timer{}
+	for op := range d.scheduleOps {
```

Comment:
Will this guarantee that the channel dequeues items are they are received? I'm wondering if there is a possibility that Unschedule is called followed by Dispatch but they are executed in the incorrect order?

Reply:
From my understanding, yes: https://gobyexample.com/range-over-channels